File: Server_Root/Server/RoomServer/MasterData/MasterDataManager.cpp
#include "stdafx.h"
#include "MasterDataManager.h"
#include <typeinfo>
#include "IMasterData.h"
#include "../Util.h"

MasterDataManager::MasterDataManager()
{
}

MasterDataManager::~MasterDataManager()
{
	for (map<string, map<int, IMasterData*>>::iterator it = m_mapData.begin(); it != m_mapData.end(); ++it)
	{
		for (map<int, IMasterData*>::iterator iter = it->second.begin(); iter != it->second.end(); ++it)
		{
			delete iter->second;
		}
		it->second.clear();
	}
	m_mapData.clear();
}

void MasterDataManager::DownloadMasterData(string strUrl)
{
	//	Temp.. load local files
	SetCharacter();
	SetSkill();
	SetBehavior();
	SetState();
	SetProjectile();
	SetItem();
	SetCharacterSpeedVariation();
}

void MasterDataManager::SetCharacter()
{
	string strTypeName = typeid(MasterData::Character).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("Character.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::Character* pCharacter = new MasterData::Character();
		pCharacter->SetData(vectorData[row]);

		m_mapData[strTypeName][pCharacter->m_nID] = pCharacter;
	}
}

void MasterDataManager::SetSkill()
{
	string strTypeName = typeid(MasterData::Skill).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("Skill.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::Skill* pSkill = new MasterData::Skill();
		pSkill->SetData(vectorData[row]);

		m_mapData[strTypeName][pSkill->m_nID] = pSkill;
	}
}

void MasterDataManager::SetBehavior()
{
	string strTypeName = typeid(MasterData::Behavior).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("Behavior.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::Behavior* pBehavior = new MasterData::Behavior();
		pBehavior->SetData(vectorData[row]);

		m_mapData[strTypeName][pBehavior->m_nID] = pBehavior;
	}
}

void MasterDataManager::SetState()
{
	string strTypeName = typeid(MasterData::State).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("State.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::State* pState = new MasterData::State();
		pState->SetData(vectorData[row]);

		m_mapData[strTypeName][pState->m_nID] = pState;
	}
}

void MasterDataManager::SetProjectile()
{
	string strTypeName = typeid(MasterData::Projectile).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("Projectile.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::Projectile* pProjectile = new MasterData::Projectile();
		pProjectile->SetData(vectorData[row]);

		m_mapData[strTypeName][pProjectile->m_nID] = pProjectile;
	}
}

void MasterDataManager::SetItem()
{
	string strTypeName = typeid(MasterData::Item).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("Item.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::Item* pItem = new MasterData::Item();
		pItem->SetData(vectorData[row]);

		m_mapData[strTypeName][pItem->m_nID] = pItem;
	}
}

void MasterDataManager::SetCharacterSpeedVariation()
{
	string strTypeName = typeid(MasterData::CharacterSpeedVariation).name();
	m_mapData[strTypeName] = map<int, IMasterData*>();

	vector<vector<string>> vectorData = Util::ReadCSV("CharacterSpeedVariation.csv");
	for (int row = 1; row < vectorData.size(); ++row)
	{
		MasterData::CharacterSpeedVariation* pCharacterSpeedVariation = new MasterData::CharacterSpeedVariation();
		pCharacterSpeedVariation->SetData(vectorData[row]);

		m_mapData[strTypeName][pCharacterSpeedVariation->m_nID] = pCharacterSpeedVariation;
	}
}

File: Client_Root/Client/Assets/Scripts/Common/MasterDataManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class MasterDataManager : MonoSingleton<MasterDataManager>
{
    private Dictionary<System.Type, Dictionary<int, IMasterData>> m_dicData = new Dictionary<System.Type, Dictionary<int, IMasterData>>();

    private void Start()
    {
        DontDestroyOnLoad(this);
    }

    public void DownloadMasterData(string strUrl)
    {
        //  Temp.. load local files
        SetCharacter();
        SetSkill();
        SetBehavior();
        SetState();
        SetProjectile();
        SetGameItem();
        SetGameItemEffect();
        SetMagic();
        SetMagicObject();
    }

    private void SetCharacter()
    {
        m_dicData[typeof(MasterData.Character)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/Character");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.Character character = new MasterData.Character();
            character.SetData(listData[row]);

            m_dicData[typeof(MasterData.Character)][character.m_nID] = character;
        }
    }

    private void SetSkill()
    {
        m_dicData[typeof(MasterData.Skill)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/Skill");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.Skill skill = new MasterData.Skill();
            skill.SetData(listData[row]);

            m_dicData[typeof(MasterData.Skill)][skill.m_nID] = skill;
        }
    }

    private void SetBehavior()
    {
        m_dicData[typeof(MasterData.Behavior)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/Behavior");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.Behavior behavior = new MasterData.Behavior();
            behavior.SetData(listData[row]);

            m_dicData[typeof(MasterData.Behavior)][behavior.m_nID] = behavior;
        }
    }

    private void SetState()
    {
        m_dicData[typeof(MasterData.State)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/State");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.State state = new MasterData.State();
            state.SetData(listData[row]);

            m_dicData[typeof(MasterData.State)][state.m_nID] = state;
        }
    }

    private void SetProjectile()
    {
        m_dicData[typeof(MasterData.Projectile)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/Projectile");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.Projectile projectile = new MasterData.Projectile();
            projectile.SetData(listData[row]);

            m_dicData[typeof(MasterData.Projectile)][projectile.m_nID] = projectile;
        }
    }

    private void SetGameItem()
    {
        m_dicData[typeof(MasterData.GameItem)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/GameItem");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.GameItem gameItem = new MasterData.GameItem();
            gameItem.SetData(listData[row]);

            m_dicData[typeof(MasterData.GameItem)][gameItem.m_nID] = gameItem;
        }
    }

    private void SetGameItemEffect()
    {
        m_dicData[typeof(MasterData.GameItemEffect)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/GameItemEffect");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.GameItemEffect gameItemEffect = new MasterData.GameItemEffect();
            gameItemEffect.SetData(listData[row]);

            m_dicData[typeof(MasterData.GameItemEffect)][gameItemEffect.m_nID] = gameItemEffect;
        }
    }

    private void SetMagic()
    {
        m_dicData[typeof(MasterData.Magic)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/Magic");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.Magic magic = new MasterData.Magic();
            magic.SetData(listData[row]);

            m_dicData[typeof(MasterData.Magic)][magic.m_nID] = magic;
        }
    }

    private void SetMagicObject()
    {
        m_dicData[typeof(MasterData.MagicObject)] = new Dictionary<int, IMasterData>();

        List<List<string>> listData = Util.ReadCSV("MasterData/MagicObject");
        for (int row = 1; row < listData.Count; ++row)
        {
            MasterData.MagicObject magicObject = new MasterData.MagicObject();
            magicObject.SetData(listData[row]);

            m_dicData[typeof(MasterData.MagicObject)][magicObject.m_nID] = magicObject;
        }
    }

    public bool GetData<T>(int nID, ref T data) where T : IMasterData
    {
        System.Type type = data == null ? typeof(T) : data.GetType();

        Dictionary<int, IMasterData> dicvalue = null;

        if (m_dicData.TryGetValue(type, out dicvalue))
        {
            IMasterData value = null;
            if (dicvalue.TryGetValue(nID, out value))
            {
                data = (T)value;

                return true;
            }
        }

        return false;
    }
}